### ISSUE TITLE
On Git Bash for Windows, poetry fails to print anything because of UnicodeEncodeError

### ISSUE THREAD

**Issue by jaepil-choi** [python-poetry/poetry#1887]
- [x] I am on the [latest](https://github.com/python-poetry/poetry/releases/latest) Poetry version.
- [x] I have searched the [issues](https://github.com/python-poetry/poetry/issues) of this repo and believe that this is not a duplicate.
- [x] If an exception occurs when executing a command, I executed it again in debug mode (`-vvv` option).

- **OS version and name**: 
```
$ uname --all
MINGW64_NT-10.0-18363 Jaepil-PC 3.0.7-338.x86_64 2019-11-21 23:07 UTC x86_64 Msys
```
I'm on Windows 10, using Git Bash for Windows. 
```
$ git --version
git version 2.25.0.windows.1
```

- **Poetry version**: Poetry version 1.0.2
- **Link of a [Gist](https://gist.github.com/) with the contents of your pyproject.toml file**: https://gist.github.com/jaepil-choi/b6ce11194f564fd179a3b43b000e236c

## Issue

This is what I get when I try to print poetry's commands with `$ poetry`

On [this issue](https://github.com/python-poetry/poetry/issues/1427), someone pointed out that this is an issue with the locale but it never worked however I changed locale. 

Also, this error is also referring to `'\xa0' in position 30` like the previous issue, so I believe it might be a poetry-specific problem. I looked up what the heck '\xa0' is, and [it is a no-break space](https://codepoints.net/U+00a0) and apparently, [it has caused a lot of problems, especially in Python](https://www.google.com/search?newwindow=1&rlz=1C1ASRM_koKR760KR760&sxsrf=ACYBGNRhPDI45k_6NEFkawqh5rePftQNjQ%3A1579019928558&ei=mO4dXtHcIdvT-Qbi6p7QBw&q=unicode+xa0+error&oq=unicode+xa0+error&gs_l=psy-ab.3...243352.244558..244672...2.0..2.344.1571.0j8j1j1......0....1..gws-wiz.......35i39j0i7i30j0i30j0i8i30j33i160j33i21.rx2joOjn5TQ&ved=0ahUKEwjRiL3Pw4PnAhXbad4KHWK1B3oQ4dUDCAs&uact=5).

I think there's a room for improvement if this error could be handled from Poetry, even though Poetry might not be directly responsible for the issue. 

Thanks for making a great package manager.  

```
Jaepil@Jaepil-PC MINGW64 /e/VSCodeProjects/poetry-demo
$ poetry -vvv
Poetry version 1.0.2

USAGE

[UnicodeEncodeError]
'cp949' codec can't encode character '\xa0' in position 30: illegal multibyte sequence

Traceback (most recent call last):
  File "C:\Users\Jaepil\.poetry\lib\poetry\_vendor\py3.7\clikit\console_application.py", line 131, in run
    status_code = command.handle(parsed_args, io)
  File "C:\Users\Jaepil\.poetry\lib\poetry\_vendor\py3.7\clikit\api\command\command.py", line 120, in handle
    status_code = self._do_handle(args, io)
  File "C:\Users\Jaepil\.poetry\lib\poetry\_vendor\py3.7\clikit\api\command\command.py", line 171, in _do_handle
    return getattr(handler, handler_method)(args, io, self)
  File "C:\Users\Jaepil\.poetry\lib\poetry\_vendor\py3.7\clikit\handler\help\help_text_handler.py", line 29, in handle
    usage.render(io)
  File "C:\Users\Jaepil\.poetry\lib\poetry\_vendor\py3.7\clikit\ui\help\abstract_help.py", line 31, in render
    layout.render(io, indentation)
  File "C:\Users\Jaepil\.poetry\lib\poetry\_vendor\py3.7\clikit\ui\layout\block_layout.py", line 42, in render
    element.render(io, self._indentations[i] + indentation)
  File "C:\Users\Jaepil\.poetry\lib\poetry\_vendor\py3.7\clikit\ui\components\labeled_paragraph.py", line 70, in render
    + '
'"
  File "C:\Users\Jaepil\.poetry\lib\poetry\_vendor\py3.7\cleo\io\io_mixin.py", line 55, in write
    super(IOMixin, self).write(string, flags)
  File "C:\Users\Jaepil\.poetry\lib\poetry\_vendor\py3.7\clikit\api\io\io.py", line 58, in write
    self._output.write(string, flags=flags)
  File "C:\Users\Jaepil\.poetry\lib\poetry\_vendor\py3.7\clikit\api\io\output.py", line 61, in write
    self._stream.write(to_str(formatted))
  File "C:\Users\Jaepil\.poetry\lib\poetry\_vendor\py3.7\clikit\io\output_stream\stream_output_stream.py", line 24, in write
    self._stream.write(string)
```

**Comment by jaepil-choi** [comment]
Googled this issue, and I think it can be handled pretty easily if encoding is set to 'utf-8' correctly. I'm not guru enough to contribute directly to poetry branch, so I'll share some findings I searched. 

example 1: https://newtoynt.tistory.com/entry/error-UnicodeEncodeError-cp949-codec-cant-encode-character-u2764-in-position-19-illegal-multibyte-sequence?category=541890

example 2: http://blog.naver.com/PostView.nhn?blogId=haya14&logNo=220781798607&parentCategoryNo=&categoryNo=20&viewDate=&isShowPopularPosts=false&from=postView

I think encoding='utf-8' should be added somewhere in the source code.

**Comment by Jarmos-san** [comment]
Not sure if the issue I stumbled across is related, if it's not, I can move it to where it'll be relevant. But I get an output on Git Bash for the command `poetry -h` which is very illegible as well.

This is what I got:
```shell
Poetry version ←[36m1.0.10←[0m

←[1mUSAGE←[0m
  ←[4mpoetry←[0m [-h] [-q] [-v [<...>]] [-V] [--ansi] [--no-ansi] [-n] <command> [<arg1>] ... [<argN>]

←[1mARGUMENTS←[0m
  ←[36m<command>←[0m              The command to execute
  ←[36m<arg>←[0m                  The arguments of the command

←[1mGLOBAL OPTIONS←[0m
  ←[36m-h←[0m (--help)            Display this help message
  ←[36m-q←[0m (--quiet)           Do not output any message
  ←[36m-v←[0m (--verbose)         Increase the verbosity of messages: "-v" for normal output, "-vv" for more verbose output and "-vvv" for debug
  ←[36m-V←[0m (--version)         Display this application version
  ←[36m--ansi←[0m                 Force ANSI output
  ←[36m--no-ansi←[0m              Disable ANSI output
  ←[36m-n←[0m (--no-interaction)  Do not ask any interactive question

←[1mAVAILABLE COMMANDS←[0m
  ←[36mabout←[0m                  Shows information about Poetry.
  ←[36madd←[0m                    Adds a new dependency to ←[32mpyproject.toml←[0m.
  ←[36mbuild←[0m                  Builds a package, as a tarball and a wheel by default.
  ←[36mcache←[0m                  Interact with Poetry's cache
  ←[36mcheck←[0m                  Checks the validity of the ←[32mpyproject.toml←[0m file.
  ←[36mconfig←[0m                 Manages configuration settings.
  ←[36mdebug←[0m                  Debug various elements of Poetry.
  ←[36menv←[0m                    Interact with Poetry's project environments.
  ←[36mexport←[0m                 Exports the lock file to alternative formats.
  ←[36mhelp←[0m                   Display the manual of a command
  ←[36minit←[0m                   Creates a basic ←[32mpyproject.toml←[0m file in the current directory.
  ←[36minstall←[0m                Installs the project dependencies.
  ←[36mlock←[0m                   Locks the project dependencies.
  ←[36mnew←[0m                    Creates a new Python project at <path>.
  ←[36mpublish←[0m                Publishes a package to a remote repository.
  ←[36mremove←[0m                 Removes a package from the project dependencies.
  ←[36mrun←[0m                    Runs a command in the appropriate environment.
  ←[36msearch←[0m                 Searches for packages on remote repositories.
  ←[36mself←[0m                   Interact with Poetry directly.
  ←[36mshell←[0m                  Spawns a shell within the virtual environment.
  ←[36mshow←[0m                   Shows information about packages.
  ←[36mupdate←[0m                 Update the dependencies as according to the ←[32mpyproject.toml←[0m file.
  ←[36mversion←[0m                Shows the version of the project or bumps it when a valid bump rule is provided.
```

**OS Version & Other Info**:

Git - git version 2.27.0.windows.1
Windows - Windows 10 Pro, version 2004, build 19041.450

**Comment by jaepil-choi** [comment]
This issue has not given much interest since I posted so I'll close the issue. 

Here's one workaround: Don't use git on windows bash. 

If you use other terminals like powershell, cmd, etc it will print help info correctly. The problem is likely due to git's behavior, not poetry's fault.

**Comment by Jarmos-san** [comment]
My suggestion for a workaround would be to use Windows Terminal instead. So my guess is the issue is with the Git Bash for Windows client itself.

**Comment by Jarmos-san** [comment]
Quoting the root cause of the problem provided [here](https://github.com/git-for-windows/git/issues/2830).

>As suspected, a non-MSYS2 `python` and `poetry` assume `ANSI` compatibility based on `TERM`. you should be able to get non-garbled output from poetry using `poetry -h --no-ansi`.

It doesn't appear to be an issue on `Git Bash for Window`'s end. I don't think the `UnicodeEncodeError` is an intended behavior, hence keeping the thread open is a good idea. Else the bug will go unnoticed even though fixing it is a low priority.

Regardless, **a better workaround**, if the output bothers you in any way, would be to use Git Bash with [Windows Terminal](https://github.com/microsoft/terminal) or use this command `poetry -h --no-ansi`

**Comment by jaepil-choi** [comment]
@Jarmos-san , thought it was better to unclog the unresolved issue for the maintainers but in your perspective, makes sense to keep it open. 



> Googled this issue, and I think it can be handled pretty easily if encoding is set to 'utf-8' correctly. I'm not guru enough to contribute directly to poetry branch, so I'll share some findings I searched.
> 
> example 1: https://newtoynt.tistory.com/entry/error-UnicodeEncodeError-cp949-codec-cant-encode-character-u2764-in-position-19-illegal-multibyte-sequence?category=541890
> 
> example 2: http://blog.naver.com/PostView.nhn?blogId=haya14&logNo=220781798607&parentCategoryNo=&categoryNo=20&viewDate=&isShowPopularPosts=false&from=postView
> 
> I think encoding='utf-8' should be added somewhere in the source code.

My initial guess was that its encoding should be changed in `io.open(foo, 'r', encoding='utf-8')` somewhere in the code but it's only a guess and the core cause might be placed at somewhere else.

**Comment by Jarmos-san** [comment]
>@Jarmos-san , thought it was better to unclog the unresolved issue for the maintainers but in your perspective, makes sense to keep it open.

I thought the same initially but after receiving confirmation from the maintainers at `Git Bash for Windows`, I think it's better to keep this thread open. I'm a bit caught up with other work, else I myself would take a look at it.

Anyway, keeping it open will be a good motivation for me to come back to it at some point in time. Besides, I don't think you clogging up the unresolved issues, it's a trivial bug that'll be fixed someday.

**Comment by jaepil-choi** [comment]
@Jarmos-san  I happened to revisit the issue after encountering `UnicodeDecodeError: 'cp949' codec can't decode byte 0xe2 in position 9921: illegal multibyte sequence` issue again. 

This time I tried to debug it and as I expected, simply adding `encoding='utf-8'` option when opening the file seems to resolve the issue. 

I'm thinking of applying the fix and make a PR. 
However, I just want to listen to another person's opinion whether making this change is a good idea. 
I don't foresee any technical problem in changing the encoding type but I'm only working on Windows platform on my local machine and I'm not so familiar with poetry internals so there might be a possible unwanted side-effect. 

Could you give me your opinion on whether this is a quick & easy & **good** fix? 

```
  File "C:\Users\chlje\.poetry\lib\poetry\_vendor\py3.8\crashtest\frame.py", line 51, in file_content
    file_content = f.read()
UnicodeDecodeError: 'cp949' codec can't decode byte 0xe2 in position 9921: illegal multibyte sequence
```

```python
    @property
    def file_content(self) -> str:
        if self._file_content is None:
            if not self._filename:
                file_content = ""
            else:
                if self._filename not in self.__class__._content_cache:
                    try:
                        with open(self._filename) as f: # change to open(self._filename, encoding='utf-8')
                            file_content = f.read()
                    except OSError:
                        file_content = ""
```

**Comment by Jarmos-san** [comment]
> Could you give me your opinion on whether this is a quick & easy & good fix?

I'm not the right person to review this as I'm not well acquainted with Poetry internals either. So, I suggest you open the PR & hope it gets reviewed & merged. If it does get merged, well that's good news for Git Bash users.

And to other Git Bash users, if this issue is bothering you, just use a PowerShell environment. I've ditched Git Bash completely (and don't even use WSL) & I couldn't be more satisfied with Poetry's support for Windows machines.

**Comment by jaepil-choi** [comment]
@Jarmos-san 

Little update. 
It turned out this `cp949` encoding issue was actually an issue from one of `poetry`'s dependency, `crashtest`. 
I made a PR on `crashtest` which will also be reflected on `poetry` once the PR is merged. 
Cheers.

**Comment by Jarmos-san** [comment]
> It turned out this `cp949` encoding issue was actually an issue from one of `poetry`'s dependency, `crashtest`.
> I made a PR([sdispater/crashtest#5](https://github.com/sdispater/crashtest/issues/5)) on `crashtest` which will also be reflected on `poetry` once the PR is merged.

Wow...I'm surprised how it turned out to be a very simple issue at the end of the day. This exactly is a prime example of how very seemingly subtle bugs can creep into production code without anyone's notice 😆 

Anyway, I look forward to seeing the PR being merged sometime soon.

**Comment by jnoelvictorino** [comment]
I tried to add a UTF-8 system locale. Might work on some.

**Comment by neersighted** [comment]
The fix is in `crashtest` 0.4.0, which is pulled in by Cleo, so we would need a Cleo release, I think -- paging @Secrus

**Comment by Secrus** [comment]
I think this should be closed now since both `crashtest` and `cleo` got released with the fix. If anyone still experiencing this issue, please open a new issue with reproduction.
